File: 340-longest-substring-with-at-most-k-distinct-characters/340-longest-substring-with-at-most-k-distinct-characters.java
class Solution {
    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        Map<Character, Integer> map = new HashMap<>();
        int l = 0;
        int n = s.length();
        int res = 0;
        for(int i = 0; i < n; i++) {
            char c1 = s.charAt(i);
            map.put(c1, map.getOrDefault(c1, 0) + 1);
            while(map.size() > k) {
                char c2 = s.charAt(l);
                map.put(c2, map.get(c2) - 1);
                if(map.get(c2) == 0) map.remove(c2);
                l++;
            }
            res = Math.max(res, i - l + 1);
        }
        return res;
    }
}

File: 340-longest-substring-with-at-most-k-distinct-characters/340-longest-substring-with-at-most-k-distinct-characters.py
class Solution:
    def lengthOfLongestSubstringKDistinct(self, s: str, k: int) -> int:
        dic = defaultdict(int)
        l, n, res = 0, len(s), 0
        
        for i in range(n):
            c1 = s[i]
            dic[c1] += 1
            while len(dic) > k:
                c2 = s[l]
                dic[c2] -= 1
                l += 1
                if dic[c2] == 0:
                    dic.pop(c2)
            res = max(res, i - l + 1)
        return res